### PR TITLE
Enable running example or creating applications

### PR DESCRIPTION
**Pull request by iddev5**
**Design**: Added a struct App in build.zig, applications can use the struct to bootstrap all the boring work of making a sysjs application. What this struct does it: create a index.html file and copy all the necessary js files into zig-out/www. The template of index.html is present in www/.

Along with this, a new file, www/loader.js is added. It defines how to actually load the wasm module and how to execute it. The path to this file would be selectable in build.zig by the user (though that feature hasn't been implemented yet).

As of this PR, running the application with an web server hasn't been implemented yet. So users can use whatever web server they want, like ``python3 -m http.server``

---

- [x] By selecting this checkbox, I agree to license my contributions to this project under the license(s) described in the LICENSE file, and I have the right to do so or have received permission to do so by an employer or client I am producing work for whom has this right.

### CODE REVIEW

File: www/index.html [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    <meta charset="utf-8">
+    <title>{[app_name]s}</title>
+    <link rel="icon" type="image/x-icon" href="favicon.ico" />
+</head>
+
+<body>
+    <script type="module">
+        {[imports]s}
+
+        import * as loader from "./loader.js";
```
Comment (emidoots) [inline www/index.html:14]: I'm not sure I see the point/value of having a separate `loader.js` - why not just inline that code here? Especially given how small it is, it seems like that would be simpler.

What am I missing?

File: src/main.zig [hunk 2]
```diff
@@ -61,7 +61,8 @@ fn Generator(comptime ZigWriter: type, comptime JSWriter: type) type {
 
         fn generateJsScaffold(gen: *@This()) !void {
             try gen.js.writeAll(
-                \\const wasmMemoryBuf = undefined;
+                \\let wasm = undefined;
+                \\let wasmMemoryBuf = undefined;
```
Comment (emidoots) [inline src/main.zig:65]: I think it'd be good to stick to Zig naming conventions even in our JS code perhaps; e.g. `wasm_memory_buf`?

File: example/main.zig [hunk 3]
```diff
@@ -3,4 +3,12 @@
 // var _ = js.doPrint;
 
 // This does:
-usingnamespace @import("sysjs_generated.zig");
+const js = @import("sysjs_generated.zig");
```
Comment (emidoots) [inline example/main.zig:6]: I think you can remove all the comments at the top of this file now

File: build.zig [hunk 4]
```diff
@@ -22,23 +21,114 @@ pub fn build(b: *std.Build) !void {
...
+            .name = options.name,
+            .compile = lib,
+            .install = install,
+        };
+    }
+
+    pub const Dependency = struct {
+        name: []const u8,
+        binding_file: []const u8,
+        wrapper_file: []const u8,
```
Comment (emidoots) [inline build.zig:128]: "binding" and "wrapper" are almost equal words for me, so it's impossible for me to tell from the names which of these is supposed to be which.

Consider using the same naming as in `generateFiles` instead:

```
    generated_sysjs_zig: []const u8,
    generated_sysjs_js: []const u8,
```

File: build.zig [hunk 5]
```diff
@@ -8,7 +8,6 @@ pub fn build(b: *std.Build) !void {
     builder_instance = b;
```
Comment (emidoots) [inline build.zig:8]: Didn't realize we had this, we should get rid of it and require the user pass in `b: *std.Build` parameter everywhere. Do we even need `pub fn getJSPath() []u8` in this file anymore?

File: build.zig [hunk 6]
```diff
@@ -22,23 +21,114 @@ pub fn build(b: *std.Build) !void {
...
 
-    const example_install = b.addInstallArtifact(lib);
     var example_compile_step = b.step("example", "Compile example");
-    example_compile_step.dependOn(&example_install.step);
+    example_compile_step.dependOn(&app.install.step);
 
-    const test_step = b.step("test", "Run library tests");
-    test_step.dependOn(&testStep(b, optimize, target).step);
+    // const test_step = b.step("test", "Run library tests");
+    // test_step.dependOn(&testStep(b, optimize, target).step);
```
Comment (emidoots) [inline build.zig:39]: Maybe just remove `testStep` and these lines entirely for now; `testStep` function is an anti-pattern I've been trying to eliminate from our codebases (prefer just writing that code directly in `pub fn build`)

**Comment by emidoots** [comment]
> As of this PR, running the application with an web server hasn't been implemented yet. So users can use whatever web server they want, like `python3 -m http.server`

I think we should keep it this way, and if folks want a nicer development web server they could use `mach build --serve` (which could work even if you're not using Mach / just using sysjs)?

File: www/index.html (hunk 1 above)
Comment (iddev5) [reply www/index.html:14]: The idea is that user can specify their custom loader.js. That way they can decide what entry point to use for wasm applications. For example, loader.init by default just calls a main function but one may overload loader.js to use a mach style entry point: separate init, update and deinit

File: src/main.zig (hunk 2 above)
Comment (iddev5) [reply src/main.zig:65]: Agreed

File: build.zig (hunk 4 above)
Comment (iddev5) [reply build.zig:128]: Completely agree

File: build.zig (hunk 5 above)
Comment (iddev5) [reply build.zig:8]: We don't need getJsPath() anymore

**Comment by iddev5** [comment]
> I think we should keep it this way, and if folks want a nicer development web server they could use mach build --serve

This is what I was thinking too

**Pull request merged**